net/dns/wsl_windows.go

```diff
@@ -0,0 +1,215 @@
+	//	C:\tsdev>wsl.exe -l
+	//	Windows Subsystem for Linux Distributions:
+	//	Ubuntu-20.04 (Default)
+	//
+	// We can skip it by passing '-q', but here we put it to work.
+	// It turns out wsl.exe -l is broken, and outputs UTF-16 names
+	// that nothing can read. (Try `wsl.exe -l | more`.)
+	// So we look at the header to see if it's UTF-16.
+	// If so, we run the rest through a UTF-16 parser.
+	//
+	// https://github.com/microsoft/WSL/issues/4607
+	var output string
+	if bytes.HasPrefix(b, []byte("W\x00i\x00n\x00d\x00o\x00w\x00s\x00")) {
+		output, err = decodeUTF16(b)
+		if err != nil {
+			logf("failed to decode wsl.exe -l output %q: %v", b, err)
+			return nil
+		}
+	} else {
+		output = string(b)
+	}
+	fmt.Printf("wslDistros: %q\n", output)
+	lines := strings.Split(output, "\n")
+	if len(lines) < 1 {
+		return nil
+	}
+	lines = lines[1:] // drop "Windows Subsystem For Linux" header
+
+	var distros []string
+	for _, name := range lines {
+		name = strings.TrimSpace(name)
+		name = strings.TrimSuffix(name, " (Default)")
+		if name == "" {
+			continue
+		}
+		fmt.Printf("wslDistros: name=%q\n", name)
+		distros = append(distros, name)
+	}
+	return distros
+}
+
+func decodeUTF16(b []byte) (string, error) {
+	if len(b) == 0 {
+		return "", nil
+	} else if len(b)%2 != 0 {
+		return "", fmt.Errorf("decodeUTF16: invalid length %d", len(b))
+	}
+	var u16 []uint16
+	for i := 0; i < len(b); i += 2 {
+		u16 = append(u16, uint16(b[i])+(uint16(b[i+1])<<8))
+	}
+	return string(utf16.Decode(u16)), nil
+}
+
+// wslManager is a DNS manager for WSL2 linux distributions.
+// It configures /etc/wsl.conf and /etc/resolv.conf.
+type wslManager struct {
+	logf     logger.Logf
+	managers map[string]directManager // distro name -> manager
+}
+
+func newWSLManager(logf logger.Logf, distros []string) *wslManager {
+	m := &wslManager{
+		logf:     logf,
+		managers: make(map[string]directManager),
+	}
+	for _, distro := range distros {
+		m.managers[distro] = newDirectManagerOnFS(wslFS{
+			user:   "root",
+			distro: distro,
+		})
+	}
+	return m
+}
+
+func (wm *wslManager) SetDNS(cfg OSConfig) error {
+	if !cfg.IsZero() {
+		if wm.setWSLConf() {
+			// What's this? So glad you asked.
+			//
+			// WSL2 writes the /etc/resolv.conf.
+			// It is aggressive about it. Every time you execute wsl.exe,
+			// it writes it. (Opening a terminal is done by running wsl.exe.)
+			// You can turn this off using /etc/wsl.conf! But: this wsl.conf
+			// file is only parsed when the VM boots up. To do that, we
+			// have to shut down WSL2.
+			//
+			// So we do it here, before we call wsl.exe to write resolv.conf.
+			if b, err := wslCommand("--shutdown").CombinedOutput(); err != nil {
+				wm.logf("WSL SetDNS shutdown: %v: %s", err, b)
+			}
+		}
+	}
+
+	for distro, m := range wm.managers {
+		if err := m.SetDNS(cfg); err != nil {
+			wm.logf("WSL(%q) SetDNS: %v", distro, err)
+		}
+	}
+	return nil
+}
+
+const wslConf = "/etc/wsl.conf"
+const wslConfSection = `# added by tailscale
```

Comment:
hmm, it would probably be good to delete this on uninstall, although that might be tricky if it's edited after; deleting the `generateResolvConf` line while leaving the section name is probably required

Reply:
It would, but it's very problematic. To touch this file, you need to call `wsl.exe --shutdown` and restart it. On my brand-new machine, that's a multiple second operation that could be disruptive. I'd like to do it zero times.

Open to ideas here, but doing this regularly doesn't seem like a great move.

Reply:
If tagged with a special comment, maybe the `generateResolvConf` line could be deleted by the NSIS uninstaller? That should be sufficiently infrequent.

Reply:
That would be infrequent enough for me, though it's a bit of work to make that happen. Would you be fine if I file a bug and do that later?

I suspect it isn't super important because it appears the nature of the generated resolv.conf has shifted over time at MS. The resolv.conf they generate points at a stub resolver on a fixed IP. Nothing I do to my computer's DNS settings change its contents. So I suspect the current behavior of this commit, restoring the backup file, will be adequate.

(That said, we should try and restore the backup on uninstall too.)

Reply:
That sounds like a plan